### PR TITLE
Physical quadratic falloff

### PR DESCRIPTION
**Pull request by bhouston**
This implements the UE4 physical light falloff equation for Spot and Point lights as proposed here: https://github.com/mrdoob/three.js/issues/6243.

I've added it via new boolean option on Point and Spot light called `physicalQuadraticFalloff`.  (In UE4, they call it "PhysicalFalloffEnabled."

In implementing the shader, I reused the spotLightDecay/pointLightDecay -- and I set it to -1 in WebGLRenderer to indicate that the light in glsl should act as physicalQuadraticFalloff.

When in this mode, light intensity is scale dependent, which means they are tricky to use, but this is how lights are in reality.  You need to get the right intensity for the scale you want to light.

I made light distance work as in UE4 when light distance > 0, and I make the light have infinite effect if light distance is 0 or less.  Thus I support both infinite lights and lights with distance-based cutoffs.

### CODE REVIEW

**Comment by mrdoob** [comment]
How about naming it `physicalFalloff`?

**Comment by bhouston** [comment]
This should be ready to merge in.  I've updated the editor to expose this functionality and it also loads and saves properly.  It should also be perfectly backwards compatible.

**Comment by WestLangley** [comment]
How about updating the docs with a _clear_ explanation that, as you said previously, it

> causes the light to have a quadratic falloff in terms of area and it assumes that the intensity of the light is in lumens and the distance is in meters.

What would be reasonable `PointLight` intensity values?

**Comment by bhouston** [comment]
@WestLangley Good idea.  I'm updated the docs and added example lumen values.

File: docs/api/lights/PointLight.html [hunk 1]
```diff
@@ -31,6 +31,8 @@
 		[page:Integer hex] — Numeric value of the RGB component of the color. <br />
 		[page:Float intensity] — Numeric value of the light's strength/intensity. <br />
 		[page:Number distance] -- The distance of the light where the intensity is 0. When distance is 0, then the distance is endless. 
+		[page:Number Decay] -- The rate of exponential decay of the list over distance.  0 is no decay, 1 is linear decay, and 2 is quadratic.
```
Comment (WestLangley) [inline docs/api/lights/PointLight.html]: light

File: docs/api/lights/SpotLight.html [hunk 2]
```diff
@@ -55,7 +55,8 @@
 		[page:Float distance] -- Maximum distance from origin where light will shine whose intensity is attenuated linearly based on distance from origin. <br />
 		[page:Radians angle] -- Maximum angle of light dispersion from its direction whose upper bound is Math.PI/2.<br />
 		[page:Float exponent] -- Rapidity of the falloff of light from its target direction.<br />
-		[page:Float decay] -- The amount the light dims along the distance of the light.
+		[page:Number Decay] -- The rate of exponential decay of the list over distance.  0 is no decay, 1 is linear decay, and 2 is quadratic.
```
Comment (WestLangley) [inline docs/api/lights/SpotLight.html]: light

File: docs/api/lights/PointLight.html (hunk 1 above)
Comment (mrdoob) [inline docs/api/lights/PointLight.html]: Also, `decay` instead of `Decay`.

File: docs/api/lights/SpotLight.html (hunk 2 above)
Comment (mrdoob) [inline docs/api/lights/SpotLight.html]: Also, `decay` instead of `Decay`.

**Comment by WestLangley** [comment]
Sorry, I still have a problem with this API.

What you are saying, is that if `light.physicalFalloff` is `true`, then the `light.decay` parameter is ignored, and the `light.intensity` parameter has different units -- `lumens`, instead of unitless. Also, the user's units of scale is required to be meters, rather than arbitrary, and a typical constructor would be

    new THREE.PointLight( 0xffffff, 1600 );

I think it would be better to create `PhysicalPointLight` and `PhysicalSpotLight`.

**Comment by bhouston** [comment]
@WestLangley I had described these caveats to its behavior fully before I implemented this PR and you did seem to approve here: https://github.com/mrdoob/three.js/issues/6243

**Comment by WestLangley** [comment]
@bhouston I was agreeing to the feature set, not the implementation. Plus, I have more information, now. I still support your idea of adding physical lighting.

In fact, we can also get rid of `decay`, which is now replaced by your new approach -- implemented as `PhysicalPointLight` and `PhysicalSpotLight`.

Plus, if you want to add `renderer.units` (https://github.com/mrdoob/three.js/issues/6259), the new "Physical" classes can accommodate it.

**Comment by bhouston** [comment]
@WestLangley In my initial issue #6243 that motivated this PR I described exactly the implementation I was going to use:

> boolean option on each light called 'physicalFalloff', which when enabled causes the light to have a quadratic falloff in terms of area and it assumes that the intensity of the light is in lumens and the distance is in meters.

This is exactly how it is implemented in this PR.

I'd suggest leaving Decay.  Decay is very useful for controlling light falloffs -- otherwise they are always constant or linear (which ever default we pick.)  Unreal follows the model in this PR -- either physical falloff via a boolean, or an explicit decay parameter while both are responsive to a distance/range parameter.

https://docs.unrealengine.com/latest/INT/Engine/Rendering/LightingAndShadows/LightTypes/Point/index.html

My recommendation is to follow what Unreal Engine does, it and Unity are the leading game engines with a lot of thought put into them.  This light is completely backwards compatible with what we have.

**Comment by WestLangley** [comment]
@bhouston I agree with what you want to do. I just suggested what I believe is a better API.

**Comment by bhouston** [comment]
@WestLangley How about to keep moving forward, we accept this PR and we can potentially in the future create new classes to separate things out going forward with new unit classes and Physical* classes.  This keeps it simple.  Please. :)

**Comment by WestLangley** [comment]
@bhouston What concerns me the most I explained in https://github.com/mrdoob/three.js/pull/6255#issuecomment-110465598. Of course, I will support whatever development path you and @mrdoob want to follow.

**Comment by mrdoob** [comment]
> I think it would be better to create `PhysicalPointLight` and `PhysicalSpotLight`.

I agree. And it's also not a big change.

@bhouston just a matter of creating new `PhysicalPointLight` class/file that extends `PointLight`. I believe the code you have is pretty much the same, it's just a cleaner API and doesn't really require much changes.

Maybe we can help with that.

@WestLangley so are you saying that by having `PhysicalPointLight`, `PointLight` will no longer need `decay`?

**Comment by mrdoob** [comment]
Thinking about it a bit more... makes a lot of sense:

```javascript
new THREE.PointLight( 0xffffff, 1 );
new THREE.PhysicalPointLight( 0xffffff, 1600 ); // decay can then be 2 by default
```

**Comment by WestLangley** [comment]
> @WestLangley so are you saying that by having PhysicalPointLight, PointLight will no longer need decay?

Yes, I did recommend that. We could revert to the prior model for `PointLight`, where the attenuation is either linear out to a max distance, or, if distance is zero, there is no attenuation at all.

Whatever you and @bhouston want to do is fine with me. I feel more strongly about having a separate "physical" class -- at least eventually.

**Comment by bhouston** [comment]
If this needs to be done I can do it.  I would prefer myself to find an easier way to simplify the constructors rather than creating new classes, but if that is the way we have to go we can go that way.

What parts of the code base need to be modified if there was to be separate classes?

I think we need to have these new classes:
- PhysicalSpotLight (constructor, clone and toJson)
- PhysicalPointLight (constructor, clone and toJson)
- Eventually a PhysicalAreaLight, but not yet. (constructor, clone and toJson)
- Support in exporters for the new light classes. (Object exporter, any others?)
- Support in importers for the new light classes. (ObjectLoader, others?)
- Support in WebGL Renderer for the new classes.  (I guess it has to translate it to the existing spot/point light shader uniforms -- thus the shaders will not change.)
- Support in other renderers for the new light classes?

This also means that if you export from any creation tool, and it doesn't export a Physical*Light, but rather a standard Light, you can not easily switch to Physical -- you need to delete the existing light and recreate one.

It will also be hard to switch between Physical and non-Physical light styles through the editor, you will have to delete a light and then recreate it again.

I know that in a visual editor, like the ThreeJS editor or Clara.io, my preferred interface is to have a boolean called Physical Falloff that I can easily turn on and off -- which is what Blender, and Unreal Engine do.  I know even if we make this change with Physical*Light classes we will likely just keep the Clara.io UI interface as a checkbox as I believe that is what users want.

BTW I also feel that we not introduce a "units" class, 3DS Max has it and it is a constant pain and it increases complexity.  Every other non-CAD tool that I know of doesn't have one and it makes it easier (e.g. Blender, Maya, Softimage, Unreal Engine, Unity 3D.)  One just needs to declare there is a unit standard (and I suggest SI units) and if you follow it, interchange is easier.

**Comment by bhouston** [comment]
Here are other ways of fixing the UI issue without creating new classes.  Just a list of patterns I know.

*Helper function for Physical:*
```
var myLight = THREE.SpotLight.newPhysical( myColor, myLumens );
```

*Use named parameters:*
```
var myLight = new THREE.SpotLight( { color: myColor, intensity: myLumens, attenuation: physical } );
```

*Why not always assume intensity is in lumens and distance in meters:*

We could have the intensity always be in lumens, which is a primary SI unit.  And distance always considered to be in meters.  I do not think this hurts or changes anything.  Of course if you use a non-physical falloff, intensity and distance don't translate into a physical result but that is just an artistic choice.

**Comment by WestLangley** [comment]
Using named parameters seems like a reasonable compromise to implement @bhouston's desired functionality with minimal changes, if @mrdoob agrees.

Here are my suggestions:

1. Since the "non-physical" attenuation model does not have a name, I would suggest the property be `physicalAttenuation`, and it take the value `true` or `false`.

2. Since I am not a fan of using a property for two purposes -- especially when the units are different, I'd keep `intensity` (which is unitless) and add `lumens` as an additional property.

3. I think it is fine to assume `meters` for distance when the physical model is used. For the non-physical model, units are irrelevant, since the non-physical attenuation model is unitless.

4. Adding the property `physicalAttenuation` allows the "magic number" `- 1` in the code to be removed. : - )

5. It is a design decision whether to use `#ifdefs` in the shader if `physicalAttenuation` is `true`. I'll leave that up to those who have an opinion on these things.

6. I still think `decay` should be removed for simplicity (it is not physical even with a value of 2 because the non-physical attenuation model is unitless) -- but I will not push it, as long as it is just considered as a tweakable parameter, and does not imply a physical model.

**Comment by bhouston** [comment]
Just a quick note, I'm out of time for contributing to ThreeJS for the time being.  I'll try to get back to this in the next few weeks.

**Comment by threejsworker** [comment]
The examples of this pullrequest are now built and visible in threejsworker. To view them, go to the following link: 

 http://threejsworker.com/viewpullrequest.html#6255

**Pull request closed without merging**